Internal API review for ContextMenuRequested (#1372)
MicrosoftEdge/WebView2Feedback: specs/ContextMenuRequested.md

# Background

There currently is no method using WebView2 APIs to customize the default context menu experience. Currently, the only option using WebView2 API's is to show or disable the default context menu. We have been requested by WebView2 app developers to allow for two different customization paths for context menus. The first option is to pass appropriate data to allow the app developers to create their own context menu UI and the second is to allow app developers to add and remove items from the default context menus.

# Description
We propose a new event for WebView2, CoreWebView2ContextMenuRequested that will allow developers to listen to context menus being requested by the end user in the embedded browser. When a context menu is requested in WebView2, the app developer will receive:

1. The list of default ContextMenuItem objects (contains name, ID, kind/type, Icon, Shorcut Desc, Access Key, the Command)
2. The coordinates that the right click

> 3. The source of the context selected

Author: Yep that is what it means, I will go back and make it more clear.

and have the choice to: 

1. Add or remove entries to the default context menu provided by the browser (not in this PR, added later)
2. Use their own UI to display their custom context menu (can either handle the selection on their own or return the selected option to the browser)

# Examples

## Win32 C++ Add/ Remove Entries From the Browser Menu

> 

Author: This is the other half of the API that will be pushed up later.

> 

Reviewer: Please remove the heading until actually adding the sample

                CoreWebView2ContextMenuItemCollectionIterator iterator = collection.GetIterator();
                args.Handled = true;
                ContextMenu cm = this.FindResource("ContextMenu") as ContextMenu;
                cm.Items.Clear();
                while (iterator.HasCurrent)
                {
                    CoreWebView2ContextMenuItem current = iterator.GetCurrent();
                    MenuItem newItem = new MenuItem();
                    newItem.Header = current.Name;
                    newItem.Click += (s, ex) => args.SelectedItem = current.id;
                    cm.Items.Add(newItem);
                    iterator.MoveNext();
                }
                cm.IsOpen = true;
            }
        }, null);
    };
```
# Remarks

# API Notes

# API Details
 ```cpp
    interface ICoreWebView2_4;
    interface ICoreWebView2ContextMenuItem;
    interface ICoreWebView2ContextMenuItemCollection;
    interface ICoreWebView2ContextMenuItemCollectionIterator;
    interface ICoreWebView2ContextMenuRequestedEventArgs;
    interface ICoreWebView2ContextMenuRequestedEventHandler;

    /// Defines the Context Menu Items
    [v1_enum]

>     typedef enum COREWEBVIEW2_CONTEXT_MENU_ITEM_ID {

Reviewer: Can we give this a more specific name than just Id? Usually we think of Ids as uniquely identifying particular menu items but that's not the case here. These are more like ids for built in menu items. Or the menu item kind/type rather than an id.

    
    /// Indicates the context selected (where the user right clicked)
    [v1_enum]
    typedef enum COREWEBVIEW2_SELECTION_CONTEXT {

        /// Indicates that a user right clicked on an image.

        COREWEBVIEW2_SELECTION_CONTEXT_IMAGE,

        /// Indicates that a user right clicked on a link.

        COREWEBVIEW2_SELECTION_CONTEXT_LINK,

        /// Indicates that a user right clicked on a textbox.

        COREWEBVIEW2_SELECTION_CONTEXT_TEXTBOX,

        /// Indicates that a user right clicked on selected text.

        COREWEBVIEW2_SELECTION_CONTEXT_TEXT,

        /// Indicates that a user right clicked on an area on the page.

        COREWEBVIEW2_SELECTION_CONTEXT_PAGE,

        /// Indicates that a user right clicked on a video

        COREWEBVIEW2_SELECTION_CONTEXT_VIDEO,
        
        /// Indicates that a user right clicked on a pdf

        COREWEBVIEW2_SELECTION_CONTEXT_PDF,

        /// list is not complete

    } COREWEBVIEW2_SELECTION_CONTEXT;

    /// Context Menu Items displayed by the Edge browser, holds the properties of a context menu item
    [uuid(7aed49e3-a93f-497a-811c-749c6b6b6c65), object, pointer_default(unique)]

>     interface ICoreWebView2ContextMenuItem : IUnknown {

Reviewer: Some context menu items have icons, checkboxes or radio buttons: 
![image](https://user-images.githubusercontent.com/2541336/121429488-d50b5b00-c92b-11eb-968c-1d0704fe3c3e.png)

I'm just looking at the browser default context menu. If you look at the browser extension's context menu API, are there more properties associated with context menu items to include?

Reviewer: How about the icons on menu items?

>     interface ICoreWebView2ContextMenuItem : IUnknown {

Author: Added a context menu type that will hold if the context menu is normal, checkbox, radio or submenu type. If it is a submenu type, the ContextMenuItem will hold a list of its Children context menu items.

>         /// Get the Name displayed for the Context Menu Item

Author: Correct, the text will have been translated. Added the documentation.

        [propget] HRESULT Name([out, retval] LPWSTR* name);

        /// Get the ID for the Context Menu Item
        [propget] HRESULT ID([out, retval] COREWEBVIEW2_CONTEXT_MENU_ITEM_ID* id);

>         /// Get the Shortcut for the Context Menu Item (CTRL+X for Cut)

Author: For the case of the Custom UI for context menu the app developer will only need to receive a string to display the shortcut and will not need to worry about the functionality. The handling of keyboard shortcuts is already done in the browser (cut, copy, paste, print, save etc) and they would only need the string to display to the user when building their own UI so no need for the app developer to parse it. Let me know if I misinterpreted.

        [propget] HRESULT Shortcut([out, retval] LPWSTR* short_cut);

    }

    /// Collection of ContextMenuItem objects 
    [uuid(f562a2f5-c415-45cf-b909-d4b7c1e276d3), object, pointer_default(unique)]
    interface ICoreWebView2ContextMenuItemCollection : IUnknown {
        /// Gets the iterator over the collection of context menu items from the start
        HRESULT GetIterator([out, retval] ICoreWebView2ContextMenuItemCollectionIterator ** iterator);

        /// Returns whether the specific Context Menu ID is contained in the collection

>         HRESULT Contains([in] COREWEBVIEW2_CONTEXT_MENU_ITEM_ID menuItem, [out, retval] BOOL * iterator);

Author: This API will also support inserting and removing entries to the ContextMenuItemCollection. Initially I had the Count and GetValueAtIndex methods, but went to the iterator version to support that. GetValueAtIndex could cause problems if the developer adds and removes context menu item entries while they are iterating through. Thoughts on this?

    [uuid(b9cc0ce5-4d10-4b9c-af06-802b199abc42), object, pointer_default(unique)]
    interface ICoreWebView2ContextMenuItemCollectionIterator : IUnknown {
        /// TRUE when the iterator has not run out of ContextMenuItem objects
        /// Will be FALSE if collection is empty or if iterator has gone past collection
        [propget] HRESULT HasCurrent([out, retval] BOOL * hasCurrent);

        /// Gets the current ICoreWebView2ContextMenuItem of the iterator
        HRESULT GetCurrent([out, retval] ICoreWebView2ContextMenuItem ** contextItem);

        /// Moves the iterator to the next ContextMenuItem
        HRESULT MoveNext([out, retval] BOOL * hasNext);

    }

    [uuid(76eceacb-0462-4d94-ac83-423a6793775e), object, pointer_default(unique)]
    interface ICoreWebView2_4 : ICoreWebView2_3
    {
        /// Add an event handler for the ContextMenuRequested event.
        /// ContextMenuRequested event is raied when the user right clicks in the webview
        ///
        /// The host can use their own UI to create their own context menu using
        /// the data provided in the API or can add to / remove from the default

>         /// context menu. provide a response with credentials for the authentication

Author: Yes, will be removed

        /// or cancel the request. If the host doesn't handle the event, Webview will
        /// display the original context menu.
        ///

>         HRESULT add_ContextMenuRequested(

Author: Will talk to him and follow up with what he says

>         HRESULT add_ContextMenuRequested(

Author: After talks with Brad and Cagri, it seems like ICoreWebView2 is more appropriate since the context menus will be specifically handled by the app and not any framework.

    runtimeclass CoreWebView2ContextMenuItem;
    runtimeclass CoreWebView2ContextMenuItemCollection;
    runtimeclass CoreWebView2ContextMenuItemCollectionIterator;

    enum CoreWebView2ContextMenuItemID {
        /// Custome type, will include Items brought in by browser extensions
        Custom = 0,

        /// No Command
        None = 1,

        /// Page Commands
        Back = 2,
        Forward = 3,
        Reload = 4,
        Stop = 5,
        NewWindow = 6,
        Print = 7,

        /// Clipboard Commands
        Cut = 8,
        Copy = 9,
        Paste = 10,
        PasteAsPlainText = 11,
        SelectAll = 12,

        /// Saving
        SaveLinkAs = 13,
        SaveImageAs = 14,
        
        Inspect = 15,
        ... /// list is not complete
    };

    enum CoreWebView2SelectionContext
    {
        Image = 0,
        Link = 1,
        Textbox = 2,
        Text = 3,
        Page = 4,
        Video = 5,
        PDF = 6,
        ... /// list is not complete
    };

    runtimeclass CoreWebView2ContextMenuRequestedEventArgs
    {
        Boolean Handled { get; set; }
        CoreWebView2SelectionContext Context { get; };

>         CoreWebView2ContextMenuItemCollection MenuItems { get; };

Author: Same thing as the above comment. The reason I did it this way was to allow for the collection to be added to and removed from by the developer in the second piece of this API. If you know of a good workaround for supporting an editable list in .Net/WinRT, that would definitely help my case.

>         CoreWebView2ContextMenuItemCollection MenuItems { get; };

Reviewer: In .NET the usual collection classes invalidate outstanding iterators if you modify the underlying collection so I'm not sure that helps. If we only have methods to modify the collection by index its sort of up to the end dev to be aware of how they're modifying the collection as they index over it, but I think that's fine.

It could be an interface with something like:

* Count
* GetValueAtIndex
* RemoveValueAtIndex
* AddValueAtIndex

In .NET we can't use IReadOnlyCollection but probably just ICollection. The implementation for this will be more involved than our existing IReadOnlyCollection conversion though so its mutable. Currently we can just create a new collection, fill it up to match our COM/Win32 collection, and hand off the .NET version. Its still doable of course it will just require more work.

        CoreWebView2ContextMenuItem SelectedItem { get; set; }
        Point Location { get; };
        Windows.Foundation.Deferral GetDeferral();
    };
    
    runtimeclass CoreWebView2ContextMenuItem
    {
        String Name { get; }
        String Shortcut { get; }

>         CoreWebView2ContextMenuItemID ID { get; }

Reviewer: By our naming conventions it should be 'Id' not 'ID'. If this is the result of our api generator tool let me know and we can open a bug against it.

```suggestion
        CoreWebView2ContextMenuItemId Id { get; }
```

                    wil::com_ptr<ICoreWebView2ContextMenuItem> newMenuItem;
                    webviewEnvironment->CreateContextMenuItem(
                        L"NewItem", L"Shorcut", &newMenuItem);    
                    CHECK_FAILURE(items->AddValueAtIndex(newMenuItem, menuCollectionCount));
                }
                return S_OK;
            })
            .Get(),
        &m_contextMenuRequestedToken));
```

## Win32 C++ Use Data to Display Custom Context Menu

The developer can use the data provided in the Event arguments to display a custom context menu with entries of their choice. For this case, the developer specifies Handled to be true and requests a deferral. deferral of this event should be completed when the user selects a context menu item (either the app developer will handle the case, or can return the selected option to the browser) or when they click on the screen (effectively closing the menu).

 ```cpp
    webview2->add_ContextMenuRequested(
        Callback<ICoreWebView2ContextMenuRequestedEventHandler>(
            [this](
                ICoreWebView2* sender,
                ICoreWebView2ContextMenuRequestedEventArgs* args)
            {

>                 auto showMenu = [this, args] {

Reviewer: Microsoft coding style has '{' on its own line. A few other places below
```suggestion
                auto showMenu = [this, args]
                {
```

                    wil::com_ptr<ICoreWebView2ContextMenuItemCollection> items;
                    CHECK_FAILURE(args->get_ContextMenuItems(&items));
                    args->put_Handled(true);

                    UINT menuCollectionCount;
                    CHECK_FAILURE(certificateCollection->get_Count(&menuCollectionCount));

                    HMENU hPopupMenu = CreatePopupMenu();
                    wil::com_ptr<ICoreWebView2ContextMenuItem> current;

                    for(UINT i = 0; i < menuCollectionCount; i++) {
                        CHECK_FAILURE(items->GetValueAtIndex(i, &current));
                        wil::unique_cotaskmem_string name;
                        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR desc;
                        wil::unique_cotaskmem_string shortcut;
                        CHECK_FAILURE(current->get_Name(&name));
                        CHECK_FAILURE(current->get_Descriptor(&desc));
                        CHECK_FAILURE(current->get_Shortcut(&shortcut));

>                         InsertMenu(hPopupMenu, 0, MF_BYPOSITION | MF_STRING, desc, name + "-" + shortcut);

Reviewer: Using 'desc' as the ID is a problem because the values aren't necessarily unique. The _CUSTOM value could come up many times from various extension menu items.

Author: Internally, each Context Menu entry has a unique id that is used for to execute the associated command. I changed the put_SelectedItem to use the context menu object instead of the descriptor. Each Context menu object will hold the unique id (private) to be used to invoke the command when selected.

                args.Handled = true;
                ContextMenu cm = this.FindResource("ContextMenu") as ContextMenu;
                cm.Items.Clear();
                for(int i = 0; i < menuList.Count; i ++){
                    CoreWebView2ContextMenuItem current = menuList[i];
                    MenuItem newItem = new MenuItem();
                    newItem.Header = current.Name;
                    newItem.Click += (s, ex) => args.SelectedItem = current.id;
                    cm.Items.Add(newItem);
                }
                cm.IsOpen = true;
            }
        }, null);
    };
```
# Remarks

# API Notes

# API Details
 ```cpp
    interface ICoreWebView2_4;
    interface ICoreWebView2Environment;
    interface ICoreWebView2ContextMenuItem;
    interface ICoreWebView2ContextMenuItemCollection;
    interface ICoreWebView2ContextMenuRequestedEventArgs;
    interface ICoreWebView2ContextMenuRequestedEventHandler;

    /// Defines the Context Menu Items
    [v1_enum]
    typedef enum COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR {
        /// Custome type, will include Items brought in by browser extensions
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_CUSTOM,

>         COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_MISSPELLED,

Reviewer: All of the enum values need /// documentation comments.

        /// No Command

>         COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_NONE,

Reviewer: What does this mean? When would it be used?

Author: Will not be used now, ContextMenuItems with no functionality (ie Submenu, separator) will return 'null' for descriptor. No "None" Descriptor needed and will be removed.

>         /// Page Commands

Reviewer: We have a tool that generates our reference documentation. /// comments are reference documentation for the next thing so these sort of grouping / heading docs won't work.

        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_BACK,
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_FORWARD,
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_RELOAD,
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_STOP,
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_NEW_WINDOW,
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_PRINT,

        /// Clipboard Commands
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_CUT,
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_COPY,
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_PASTE,
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_PASTE_AS_PLAIN_TEXT,
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_SELECT_ALL,

        /// Saving
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_SAVE_LINK_AS,
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_SAVE_IMAGE_AS,

        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_INSPECT, 
        
    } COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR;
    
    /// Indicates the context selected (where the user right clicked)
    [v1_enum]

>     typedef enum COREWEBVIEW2_SELECTION_CONTEXT {

Reviewer: It seems like this is missing other contexts like links, or text. Why are some things in this list and not others?

        /// Indicates that a user right clicked on the page.

>         COREWEBVIEW2_SELECTION_CONTEXT_DEFAULT,

Reviewer: The comment says 'page' and the enum name says 'default'. Should this be _PAGE?

        /// Indicates that a user right clicked on audio.
        COREWEBVIEW2_SELECTION_CONTEXT_AUDIO,

        /// Indicates that a user right clicked on an image.
        COREWEBVIEW2_SELECTION_CONTEXT_IMAGE,
        
        /// Indicates that a user right clicked on a video
        COREWEBVIEW2_SELECTION_CONTEXT_VIDEO,

    } COREWEBVIEW2_SELECTION_CONTEXT;

    /// Indicates the menu item type
    [v1_enum]
    typedef enum COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND {
        /// A normal menu item kind, can be enabled or not.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_NORMAL,

        /// A checkbox. Can be checked or enabled
        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_CHECKBOX,

        /// A radio button type, can be checked or enabled
        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_RADIO,

>         /// A Separator type, has no underlying functionality

Reviewer: What does 'no underlying functionality mean'? Do you mean it cannot be disabled and cannot be checked?
```suggestion
        /// A separator type, has no underlying functionality
```

Author: I will change the wording to be more clear, but I meant it cannot be invoked or checked. Disabling its visibility would still be possible for the end dev in case they don't want a separator in the UI.

    {
        /// Add an event handler for the ContextMenuRequested event.
        /// ContextMenuRequested event is raied when the user right clicks in the webview
        ///
        /// The host can use their own UI to create their own context menu using
        /// the data provided in the API or can add to / remove from the default
        /// context menu. If the host doesn't handle the event, Webview will
        /// display the original context menu.

        HRESULT add_ContextMenuRequested(
            [in] ICoreWebView2ContextMenuRequestedEventHandler* eventHandler,
            [out] EventRegistrationToken* token);

        /// Remove an event handler previously added with add_ContextMenuRequested.
        HRESULT remove_ContextMenuRequested(
            [in] EventRegistrationToken token);
    }

    /// A continuation of the ICoreWebView2Environment interface.
    [uuid(04d4fe1d-ab87-42fb-a898-da241d35b63c), object, pointer_default(unique)]
    interface ICoreWebView2Environment : IUnknown {
        /// Create a ContextMenuItem object used for developers to insert new items 
        /// into the default browser context menu, Descriptor will be CUSTOM

>         HRESULT CreateContextMenuItem(

Reviewer: You have new properties on your context menu item like enabled, checked and so on that the end developer cannot modify. Do you want to have them as parameters in CreateContextMenuItem or make them settable properties on the context menu item interface?

Author: I think it makes more sense to include them as parameters in the CreateContextMenuItem since they shouldn't be settable once the context menu gets shown. Checked shouldn't be settable because it's state will change once the end user clicks on it. Enabled shouldn't be settable either because if the end developer wants to disable an item, they might as well remove it completely. Checked, enabled etc are just for displaying the custom context menu correctly and should be parameters for CreateContextMenuItem.

    

    enum CoreWebView2ContextMenuItemDescriptor {
        /// Custome type, will include Items brought in by browser extensions
        Custom = 0,

        /// No Command
        None = 1,

        /// Page Commands
        Back = 2,
        Forward = 3,
        Reload = 4,
        Stop = 5,
        NewWindow = 6,
        Print = 7,

        /// Clipboard Commands
        Cut = 8,
        Copy = 9,
        Paste = 10,
        PasteAsPlainText = 11,
        SelectAll = 12,

        /// Saving
        SaveLinkAs = 13,
        SaveImageAs = 14,
        
        Inspect = 15,
    };

    enum CoreWebView2SelectionContext
    {
        Default = 0,
        Audio = 1,
        Image = 2,
        Video = 3,
    };

    enum CoreWebView2ContextMenuItemKind
    {
        Normal = 0,
        Checkbox = 1,
        Radio = 2,
        Separator = 3,
        Submenu = 4,
    };

    runtimeclass CoreWebView2ContextMenuRequestedEventArgs
    {
        Boolean Handled { get; set; }
        CoreWebView2SelectionContext Context { get; };
        Boolean IsSourceSelected { get; }
        Boolean IsSourceEditable { get; }
        Boolean IsSourceEnabled { get; }

>         Uri ArnchorUri { get; }

Reviewer: ```suggestion
        Uri AnchorUri { get; }
```

        IVector<CoreWebView2ContextMenuItem> MenuItems { get; };
        CoreWebView2ContextMenuItem SelectedItem { get; set; }
        Point Location { get; };
        Windows.Foundation.Deferral GetDeferral();
    };
    
    runtimeclass CoreWebView2ContextMenuItem
    {
        String Name { get; }
        String Shortcut { get; }
        CoreWebView2ContextMenuItemDescriptor Descriptor { get; }
        CoreWebView2ContextMenuItemKind Kind { get; }
        Boolean IsEnabled { get; }
        Boolean IsChecked { get; }
        IVector<CoreWebView2ContextMenuItem> Children { get; }
    };

    runtimeclass CoreWebView2Environment
    {
        public CoreWebView2ContextMenuItem CreateContextMenuItem(
            String name,

>             String shortcut,

Reviewer: trailing ',' is not allowed

>         )

Reviewer: missing ;

 ```cpp
    webview2->add_ContextMenuRequested(
        Callback<ICoreWebView2ContextMenuRequestedEventHandler>(
            [this](
                ICoreWebView2* sender,
                ICoreWebView2ContextMenuRequestedEventArgs* args)
            {
                auto showMenu = [this, args]
                {
                    wil::com_ptr<ICoreWebView2ContextMenuItemCollection> items;
                    CHECK_FAILURE(args->get_MenuItems(&items));
                    args->put_Handled(true);

                    UINT32 itemsCount;
                    CHECK_FAILURE(items->get_Count(&itemsCount));

                    HMENU hPopupMenu = CreatePopupMenu();
                    wil::com_ptr<ICoreWebView2ContextMenuItem> current;

                    for (UINT32 i = 0; i < itemsCount; i++) 
                    {
                        CHECK_FAILURE(items->GetValueAtIndex(i, &current));
                        wil::unique_cotaskmem_string name;
                        CHECK_FAILURE(current->get_Name(&name));
                        UINT32 command_id;

>                         CHECK_FAILURE(current->get_CommandId(&command_id));

Author: The command ID is the actual edge-defined context menu item id that is used to invoke the command on the server side. I originally did have it where they could return the index, the problem that I faced was with the submenu items and how to use the index there. My solution was to use the command ID because it is unique and already a property of the server side context menu item, but would love to hear your thoughts.

                        InsertMenu(hPopupMenu, 0, MF_BYPOSITION | MF_STRING, command_id, name);
                    }
                    HWND hWnd;
                    m_appWindow->GetWebViewController()->get_ParentWindow(&hWnd);
                    SetForegroundWindow(hWnd);
                    wil::com_ptr<ICoreWebView2ContextMenuInfo> parameters;
                    args->get_ContextMenuInfo(&parameters);
                    POINT locationInControlCoordinates;
                    POINT locationInScreenCoordinates;
                    CHECK_FAILURE(parameters->get_Location(&locationInControlCoordinates));
                    // get_Location returns coordinates in relation to upper left Bounds of the WebView2.Controller. Will need to convert to Screen coordinates to display the popup menu in the correct location.
                    ConvertToScreenCoordinates(locationInControlCoordinates, &locationInScreenCoordinates);
                    UINT32 selectedCommand = TrackPopupMenu(hPopupMenu, TPM_TOPALIGN | TPM_LEFTALIGN | TPM_RETURNCMD, locationInScreenCoordinates.x, locationInScreenCoordinates.y, 0, hWnd, NULL);

>                     CHECK_FAILURE(args->put_SelectedCommand(selectedCommand));

Author: The reason I have them return the unique command ID here is because I thought it would be easier for the developer to return an id that was given to them rather than having to return the selected ContextMenuItem object, especially if the object selected is nested inside a few layers of submenus (would be hard to retrieve an item based off index when submenus come into play). If you feel like that is something the developer will have to work around, let me know and I can change it.

    } COREWEBVIEW2_CONTEXT_TYPE;

    /// Indicates the menu item type.
    [v1_enum]
    typedef enum COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND
    {
        /// A normal menu item kind, can be enabled or not.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_NORMAL,

        /// A checkbox. Can be checked or enabled.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_CHECKBOX,

        /// A radio button type, can be checked or enabled.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_RADIO,

        /// A separator type, used to visually differentiate sections of context menu items.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_SEPARATOR,

        /// A submenu type, will contain a list of its children menu items.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_SUBMENU,
    } COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND;

    /// Represents a context menu item of a context menu displayed by WebView.
    [uuid(7aed49e3-a93f-497a-811c-749c6b6b6c65), object, pointer_default(unique)]
    interface ICoreWebView2ContextMenuItem : IUnknown
    {
        /// Get the label displayed for the ContextMenuItem, will already have been translated by the WebView.
        [propget] HRESULT Label([out, retval] LPWSTR* value);

>         /// Get the name for the ContextMenuItem, will be an unlocalized string to allow for 

Author: The name would most likely be set for all of them but will only be in use for the extension items and other items since those are the two items that can have multiple items that fall in those enum categories. I will specify that in the description.

        /// have those descriptors.
        [propget] HRESULT Name([out, retval] LPWSTR* value);

        /// Get the descriptor for the ContextMenuItem.
        [propget] HRESULT Descriptor([out, retval] COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR* value);

        /// Get the Command ID for the context menu item, this ID is unique and will be used by the app to return which specific item was selected.
        [propget] HRESULT CommandId([out, retval] UINT32* value);

        /// Get the shortcut for the ContextMenuItem.
        [propget] HRESULT Shortcut([out, retval] LPWSTR* value);

        /// Get the Icon for the ContextMenuItem in PNG format in the form of an IStream.
        [propget] HRESULT Icon([out, retval] IStream** value);

        /// Returns the menu item kind.
        [propget] HRESULT Kind([out, retval] COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND* value);

        /// Returns if the menu item is enabled.
        [propget] HRESULT IsEnabled([out, retval] BOOL* value);

        /// Returns if the menu item is checked, used for checkbox or radio types.
        [propget] HRESULT IsChecked([out, retval] BOOL* value);

>         /// Returns the list of Children menu items if the kind is submenu.

Author: Specified that it will return null.

        /// Gets whether the `ContextMenuRequested` event is handled by host.
        [propget] HRESULT Handled([out, retval] BOOL* value);

        /// Returns an `ICoreWebView2Deferral` object. Use this operation to
        /// complete the event when a user selects a context menu item from custom menu
        /// or clicks outside from context menu.
        HRESULT GetDeferral([out, retval] ICoreWebView2Deferral ** deferral);
    }

    /// Context Menu Parameters including the location of the request, the context selected and the appropriate data associated with the context menu request.
    [uuid(b8611d99-eed6-4f3f-902c-a198502ad472), object, pointer_default(unique)]
    interface ICoreWebView2ContextMenuInfo : IUnknown
    {
        /// The coordinates where the context menu is requested occured in relation to the upper left corner of the webview bounds.
        [propget] HRESULT Location([out, retval] POINT* value);

        /// The type of context that the user selected.
        [propget] HRESULT Context([out, retval] COREWEBVIEW2_CONTEXT_TYPE* value);

        /// The url of the page.
        [propget] HRESULT PageUrl([out, retval] LPWSTR * value);

>         /// The url of the frame (if context menu invoked on a frame, null otherwise).

Author: They are not mutually exclusive. They can sometimes be null but you can definitely have multiple occurring at the same time in many cases.

        [propget] HRESULT FrameUrl([out, retval] LPWSTR * value);

        /// The source url of element (if context menu invoked on media element, null otherwise).
        [propget] HRESULT SourceUrl([out, retval] LPWSTR * value);

        /// The url of the link (if context menu invoked on a link, null otherwise).
        [propget] HRESULT LinkUrl([out, retval] LPWSTR * value);

>         /// The text of the link (if context menu invoked on a link, null otherwise).

Author: Also not mutually exclusive.

                        IID_PPV_ARGS(&webviewEnvironment));
                    wil::com_ptr<ICoreWebView2ContextMenuItem> newMenuItem;
                    webviewEnvironment->CreateContextMenuItem(
                        1, L"Display Link", L"Shorcut", nullptr, COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_NORMAL, false, &newMenuItem);    
                    CHECK_FAILURE(items->AddValueAtIndex(menuCollectionCount, newMenuItem.get()));
                }
                return S_OK;
            })
            .Get(),
        &m_contextMenuRequestedToken);
    
    /// If newly inputted item (with CreateContextMenuItem) was selected
    webview2->add_CustomItemSelected(
        Callback<ICoreWebView2CustomItemSelectedEventHandler>(
              [this](
                ICoreWebView2* sender,
                ICoreWebView2CustomItemSelectedEventArgs* args)
            {
                UINT32 customId;
                args->get_SelectedId(&customId);
                wil::com_ptr<ICoreWebView2ContextMenuParams> params;
                args->get_ContextMenuParams(&params);
                switch(customId)
                {

>                     case 1:

Author: I like this idea. Will change the code to reflect this.

        switch (customId)
        {
            case 1:
                string linkUrl = parameters.LinkUrl;
                MessageBox.Show(linkUrl, "Display Link", MessageBoxButton.YesNo);
            break;
        };
    };
``` 

## .Net/ WinRT Use Data to Display Custom Context Menu 

 ```c#
    webView.CoreWebView2.ContextMenuRequested += delegate (object sender, CoreWebView2ContextMenuRequestedEventArgs args)
    {
        CoreWebView2Deferral deferral = args.GetDeferral();
        System.Threading.SynchronizationContext.Current.Post((_) =>
        {
            using (deferral)
            {
                IList<CoreWebView2ContextMenuItem> menuList = args.MenuItems;
                CoreWebView2ContextType context = args.ContextMenuParams.context;
                args.Handled = true;
                ContextMenu cm = this.FindResource("ContextMenu") as ContextMenu;
                cm.Items.Clear();
                for(int i = 0; i < menuList.Count; i ++){
                    CoreWebView2ContextMenuItem current = menuList[i];
                    MenuItem newItem = new MenuItem();
                    newItem.Header = current.Name;

>                     newItem.Click += (s, ex) => args.SelectedItem = current;

Reviewer: Do C# lambdas capture the value of current at the time the lambda is created or the variable current so that all the lambdas end up with current as its final value?

Author: I played with this using a message box and it seems like all the 'current' values used were at the time of lambda creation. I am not sure if this test directly translates to how it will act in the above scenario but it seemed like it captured them so they didn't end up with the same value.
What I did was: 
`                    for(int i = 0; i < arr.Length; i++)
                    {
                        string current = arr[i];
                        MenuItem newOne = new MenuItem();
                        newOne.Header = current;
                        newOne.Click += (s, ex) => MessageBox.Show(current, current, MessageBoxButton.YesNo);
                        cm.Items.Add(newOne);
                    }`

        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_RADIO,

        /// A separator type, used to visually differentiate sections of context menu items. Can not be disabled or checked but can be deleted from the list.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_SEPARATOR,

        /// A submenu type, will contain a list of its children
        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_SUBMENU,

    } COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND;

    /// Context Menu Items displayed by the Edge browser, holds the properties of a context menu item
    [uuid(7aed49e3-a93f-497a-811c-749c6b6b6c65), object, pointer_default(unique)]
    interface ICoreWebView2ContextMenuItem : IUnknown
    {
        /// Get the name displayed for the Context Menu Item, will already have been translated by the browser
        [propget] HRESULT Name([out, retval] LPWSTR* value);

        /// Get the descriptor for the Context Menu Item
        [propget] HRESULT Descriptor([out, retval] COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR* value);

        /// Get the shortcut for the Context Menu Item, the functionality will already exist in the browser code
        [propget] HRESULT Shortcut([out, retval] LPWSTR* value);

>         /// Get the Icon for the ContextMenuItem in the form of IStream

Reviewer: We must also document what is the format of the image data in the IStream

Author: It seems like the Menu Model using ImageModel which uses Image and that will represent the icon in the form of a PNG. Will document that.

        [propget] HRESULT Icon([out, retval] IStream** value);

        /// Returns the menu item kind
        [propget] HRESULT Kind([out, retval] COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND* value);

>         /// Returns if the menu item is visible

Reviewer: When is this used? If an end dev is constructing context menu UI to display do they need to check if any of the menu items !IsVisible and then exclude it from the context menu? Should we just exclude them in this case? Is this from the browser extension context menu APIs?

Author: You are right. We should just exclude them in this case. I was thinking about the case where a menu item is not visible, but excluding it from the list we give the devs is definitely the correct way to go.

            [out] EventRegistrationToken* token);

        /// Remove an event handler previously added with add_ContextMenuRequested.
        HRESULT remove_ContextMenuRequested(
            [in] EventRegistrationToken token);

        /// Add an event handler for the CustomItemSelected event.
        /// CustomItemSelected event is raied when the user selects a context menu item that
        /// was inserted by the developer and is not part of the default browser context menu.
        /// The developer will include the logic for the appropriate data in this event handler.

        HRESULT add_CustomItemSelected(
            [in] ICoreWebView2CustomItemSelectedEventHandler* eventHandler,
            [out] EventRegistrationToken* token);

        /// Remove an event handler previously added with add_CustomItemSelected.
        HRESULT remove_CustomItemSelected(
            [in] EventRegistrationToken token);
        
    }

    /// A continuation of the ICoreWebView2Environment interface.
    [uuid(04d4fe1d-ab87-42fb-a898-da241d35b63c), object, pointer_default(unique)]
    interface ICoreWebView2Environment : IUnknown
    {

>         /// Create a ContextMenuItem object used for developers to insert new items into the default browser context menu

Reviewer: ```suggestion
        /// Create a ContextMenuItem object used for developers to insert new items into the default browser context menu.
```

>         /// The Enabled and visible properties will default to true and the IsChecked property will only be used if the 

Reviewer: Saying they *default* to true confused me because it sounds like they default to one value but can be changed to another. I think rather you're saying that enabled and visible are just always true for custom context menu items right?
```suggestion
        /// The IsEnabled and IsVisible properties are always true for custom context menu items. The IsChecked property will only be used if the 
```

>         /// menu item type is radio or checkbox

Reviewer: ```suggestion
        /// menu item type is radio or checkbox.
```

    interface ICoreWebView2ContextMenuRequestedEventHandler : IUnknown
    {
        /// Called to provide the event args when a context menu is invoked on WebView2 element
        HRESULT Invoke(
            [in] ICoreWebView2* sender,
            [in] ICoreWebView2ContextMenuRequestedEventArgs* args);
    }

    [uuid(49e1d0bc-fe9e-4481-b7c2-32324aa21998), object, pointer_default(unique)]
    interface ICoreWebView2CustomItemSelectedEventHandler : IUnknown
    {
        /// Called to provide the event args when the end user selects on an dev-inserted context menu item
        HRESULT Invoke(
            [in] ICoreWebView2* sender,
            [in] ICoreWebView2CustomItemSelectedEventArgs* args);
    }

    /// Event args for the ContextMenuRequested event. Will contain the selection information and a collection of all of the default context menu items that the default WebView2 menu would show and allows the app to draw its own context menu or add/ remove from the default context menu.
    [uuid(a1d309ee-c03f-11eb-8529-0242ac130003), object, pointer_default(unique)]
    interface ICoreWebView2ContextMenuRequestedEventArgs : IUnknown
    {

>         /// The list of default ContextMenuItem objects

Reviewer: You say in the docs for the event args, but it would be nice to also explicitly note in the docs for the property that you can change the collection.
```suggestion
        /// The list of ContextMenuItem objects. The end developer may modify the contents of this collection.
```

        [propget] HRESULT MenuItems([out, retval] ICoreWebView2ContextMenuItemCollection ** value);

        /// Contains the data regarding the selection

>         [propget] HRESULT ContextMenuParams([out, retval] ICoreWebView2ContextMenuParams ** value);

Reviewer: We generally avoid abbreviations. This should be ContextMenuParameters.
```suggestion
        [propget] HRESULT ContextMenuParameters([out, retval] ICoreWebView2ContextMenuParameters ** value);
```

        /// Returns the selected Context Menu Item
        [propget] HRESULT SelectedItem([out, retval] ICoreWebView2ContextMenuItem ** value);

>         /// Sets the Selected Menu Item for the browser to execute the command

Reviewer: Please note in the documentation what the default value for this property is and what it means to set this value to null. I assume this value defaults to null and that null means there was no selection and the context menu was cancelled?

        /// Context menu item descriptor for "Save link as" action.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_SAVE_LINK_AS,

        /// Context menu item descriptor for "Copy link" action.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_COPY_LINK,

        /// Context menu item descriptor for "Save image as" action.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_SAVE_IMAGE_AS,

        /// Context menu item descriptor for "Copy image" action.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_COPY_IMAGE,

        /// Context menu item descriptor for "Copy image link" action.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_COPY_IMAGE_LINK,

        /// Context menu item descriptor for "Save media as" action.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_SAVE_MEDIA_AS,

        /// Context menu item descriptor for other commands not corresponding with any of the enums above.
        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR_OTHER,
    } COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR;
    
    /// Indicates the type of context for which the context menu was created
    /// for the `ICoreWebView2ContextMenuInfo::get_ContextType` method
    [v1_enum]

>     typedef enum COREWEBVIEW2_CONTEXT_TYPE

Reviewer: We generally use the word Kind instead of Type when we're not talking about types in the sense of programming languages.

```suggestion
    typedef enum COREWEBVIEW2_CONTEXT_KIND
```

        /// Returns an `ICoreWebView2Deferral` object. Use this operation to
        /// complete the event when the custom context menu is closed.
        HRESULT GetDeferral([out, retval] ICoreWebView2Deferral** deferral);
    }

    /// Represents the information regarding the context menu. Includes the location of the request, 
    /// the context selected and the appropriate data used for the actions of a context menu.
    [uuid(b8611d99-eed6-4f3f-902c-a198502ad472), object, pointer_default(unique)]
    interface ICoreWebView2ContextMenuInfo : IUnknown
    {
        /// Gets the coordinates where the context menu request occured in relation to the upper left corner of the webview bounds.
        [propget] HRESULT Location([out, retval] POINT* value);

        /// Gets the type of context that the user selected.
        [propget] HRESULT ContextType([out, retval] COREWEBVIEW2_CONTEXT_TYPE* value);

        /// Returns TRUE if the context menu was requested on the main frame and
        /// FALSE if invoked on another frame.
        [propget] HRESULT IsMainFrame([out, retval] BOOL* value);

        /// Gets the url of the page.

>         [propget] HRESULT PageUrl([out, retval] LPWSTR* value);

Reviewer: We generally use the term URI not URL in our API. Applies to the following properties as well.
```suggestion
        [propget] HRESULT PageUri([out, retval] LPWSTR* value);
```
